File: midi_lab/visualizer/styles_builders/layout.py
# -*- coding: utf-8 -*-
"""
NDC レイアウト — Qt OpenGL 座標系に合わせる。

- X: -1（左）〜 +1（右）= ウィジェット全幅
- Y: -1（下）〜 +1（上）= 鍵盤は下端、ノートは上から落下
"""
from __future__ import annotations

from midi_lab.visualizer.styles_builders.context import RenderContext


def kb_bottom_ndc(kb_ratio: float = 0.14) -> float:
    return -1.0


def kb_top_ndc(kb_ratio: float = 0.14) -> float:
    return -1.0 + kb_ratio * 2.0


def play_line_ndc(ctx: RenderContext) -> float:
    """判定ライン — 鍵盤の上端。"""
    return kb_top_ndc(ctx.kb_ratio)


def lane_top_ndc(ctx: RenderContext) -> float:
    """ノート落下レーンの上端。"""
    return 1.0 - 0.05


def fall_span_ndc(ctx: RenderContext) -> float:
    return lane_top_ndc(ctx) - play_line_ndc(ctx)


def y_at_beat(ctx: RenderContext, beat: float) -> float:
    """拍位置 → NDC Y。過去の拍ほど下（プレイラインを通過済み）。"""
    play = play_line_ndc(ctx)
    span = max(fall_span_ndc(ctx), 0.2)
    return play - (ctx.t_ql - beat) / max(ctx.window_ql, 0.001) * span


def ndc_to_pixel_y(ndc_y: float, height: int) -> float:
    """NDC Y → ピクセル Y（Qt: 上が 0）。"""
    return (1.0 - (ndc_y + 1.0) / 2.0) * height


def ndc_to_pixel_x(ndc_x: float, width: int) -> float:
    return (ndc_x + 1.0) * 0.5 * width


def midi_to_lane_y(midi: int, y_lo: int, y_hi: int, *, margin: float = 0.08) -> float:
    """Cyber 等 — 低音は下（鍵盤付近）、高音は上。"""
    span = max(y_hi - y_lo, 1)
    t = (midi - y_lo) / span
    bot = kb_top_ndc(0.14) + margin
    top = lane_top_ndc_from_ratio(0.14) - margin
    return bot + t * (top - bot)


def play_line_ndc_from_ratio(kb_ratio: float) -> float:
    return kb_top_ndc(kb_ratio)


def lane_top_ndc_from_ratio(kb_ratio: float) -> float:
    return 1.0 - 0.05

File: rust/midi_viz/src/render/layout.rs
//! NDC レイアウト — Python `layout.py` と同一仕様（Y: 下=-1, 上=+1）

pub fn kb_bottom_ndc(kb_ratio: f32) -> f32 {
    -1.0
}

pub fn kb_top_ndc(kb_ratio: f32) -> f32 {
    -1.0 + kb_ratio * 2.0
}

pub fn play_line_ndc(kb_ratio: f32) -> f32 {
    kb_top_ndc(kb_ratio)
}

pub fn lane_top_ndc(_kb_ratio: f32) -> f32 {
    1.0 - 0.05
}

pub fn fall_span_ndc(kb_ratio: f32) -> f32 {
    lane_top_ndc(kb_ratio) - play_line_ndc(kb_ratio)
}

pub fn y_at_beat(t_ql: f64, beat: f64, window_ql: f64, kb_ratio: f32) -> f32 {
    let play = play_line_ndc(kb_ratio) as f64;
    let span = fall_span_ndc(kb_ratio).max(0.2) as f64;
    let w = window_ql.max(0.001);
    (play - (t_ql - beat) / w * span) as f32
}

pub fn midi_to_theta(midi: u8, y_lo: u8, y_hi: u8) -> f32 {
    let span = (y_hi - y_lo).max(1) as f32;
    let t = (midi - y_lo) as f32 / span;
    t * std::f32::consts::TAU - std::f32::consts::FRAC_PI_2
}

pub fn polar_xy(aspect: f32, angle: f32, radius: f32) -> (f32, f32) {
    (
        angle.cos() * radius,
        angle.sin() * radius * aspect,
    )
}

pub fn x_ndc(beat: f64, x0: f64, x1: f64) -> f32 {
    let span = (x1 - x0).max(0.001);
    ((beat - x0) / span * 2.0 - 1.0) as f32
}

pub fn midi_to_ndc_y(midi: u8, y_lo: u8, y_hi: u8, kb_ratio: f32) -> f32 {
    let span = (y_hi - y_lo).max(1) as f32;
    let t = (midi - y_lo) as f32 / span;
    let margin = 0.08f32;
    let bot = play_line_ndc(kb_ratio) + margin;
    let top = lane_top_ndc(kb_ratio) - margin;
    bot + t * (top - bot)
}

pub fn build_key_x(midi: u8, y_lo: u8, y_hi: u8) -> (f32, f32) {
    let blacks = [1u8, 3, 6, 8, 10];
    let whites: Vec<u8> = (y_lo..=y_hi)
        .filter(|m| !blacks.contains(&(m % 12)))
        .collect();
    if whites.is_empty() {
        let span = (y_hi - y_lo).max(1) as f32;
        let t = (midi - y_lo) as f32 / span;
        let xc = -1.0 + t * 2.0;
        return (xc - 0.012, xc + 0.012);
    }
    if !blacks.contains(&(midi % 12)) {
        if let Some(i) = whites.iter().position(|&w| w == midi) {
            let n = whites.len() as f32;
            let xl = -1.0 + (i as f32 / n) * 2.0;
            let xr = -1.0 + ((i + 1) as f32 / n) * 2.0;
            return (xl, xr);
        }
    }
    let span = (y_hi - y_lo).max(1) as f32;
    let t = (midi - y_lo) as f32 / span;
    let xc = -1.0 + t * 2.0;
    (xc - 0.012, xc + 0.012)
}
